src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c

```diff
@@ -349,6 +385,226 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
 	return ret;
 }
 
+int xocl_create_hw_context(struct xocl_dev *xdev, struct drm_file *filp,
+		struct drm_xocl_create_hw_ctx *hw_ctx_args, uint32_t slot_id)
+{
+	struct kds_client *client = filp->driver_priv;
+	struct kds_client_hw_ctx *hw_ctx = NULL;
+	uuid_t *xclbin_id = NULL;
+	int ret = 0;
+
+	if (!client)
+		return -EINVAL;
+
+	ret = XOCL_GET_XCLBIN_ID(xdev, xclbin_id);
+	if (ret)
+		return ret;
+
+	mutex_lock(&client->lock);
+
+	hw_ctx = kds_alloc_hw_ctx(client, xclbin_id, slot_id);
+	if (!hw_ctx) {
+		ret = -EINVAL;
+		goto error_out;
+	}
+
+	/* Lock the bitstream. Unlock the same in destroy context */
+	ret = xocl_icap_lock_bitstream(xdev, xclbin_id);
+	if (ret) {
+		kds_free_hw_ctx(client, hw_ctx);
+		ret = -EINVAL;
+		goto error_out;
+	}
+
+	hw_ctx_args->hw_context = hw_ctx->hw_ctx_idx;
+
+error_out:
+	mutex_unlock(&client->lock);
+	XOCL_PUT_XCLBIN_ID(xdev);
+	return ret;
+}
+
+int xocl_destroy_hw_context(struct xocl_dev *xdev, struct drm_file *filp,
+		struct drm_xocl_destroy_hw_ctx *hw_ctx_args)
+{
+        struct kds_client_hw_ctx *hw_ctx = NULL;
+	struct kds_client *client = filp->driver_priv;
+        int ret = 0;
+
+        mutex_lock(&client->lock);
+
+        hw_ctx = kds_get_hw_ctx_by_id(client, hw_ctx_args->hw_context);
+        if (!hw_ctx) {
+                userpf_err(xdev, "No valid HW context is open");
+		mutex_unlock(&client->lock);
```

Comment:
It is best to only have 1 mutex_lock and 1 mutex_unlock in a function.  Please refactor the code.

Hint
Instead of return back a -EINVAL
1. Set the variable ret to this value
2. Add a label at line 447
3. Add a goto here and call the label on line 447.

Reply:
Fixed. Thanks